lib/ExtractSourceMapPlugin.js

```diff
@@ -0,0 +1,314 @@
+			"ExtractSourceMapPlugin",
+			(compilation, { normalModuleFactory }) => {
+				const modulesMap = new Map();
+				/**
+				 * @param {JavascriptParser} parser parser
+				 */
+				const handler = parser => {
+					parser.hooks.finish.tap("ExtractSourceMapPlugin", (_, comments) => {
+						const mapping = getSourceMappingURL(comments);
+						if (!mapping || !mapping.sourceMappingURL) return;
+						const { node, sourceMappingURL } = mapping;
+						const module = parser.state.module;
+
+						const dep = new ConstDependency(
+							"// extracted source map",
+							node.range
+						);
+						dep.loc = node.loc;
+						module.addPresentationalDependency(dep);
+
+						if (!module.useSourceMap || !("setSourceMap" in module)) return;
+						const source = module.originalSource();
+						if (!source) return;
+						if (source instanceof SourceMapSource) {
+							const warnings = new WebpackError(
+								"ExtractSourceMapPlugin: Source already has source map."
+							);
+							warnings.details =
+								"It maybe caused by using 'source-map-loader'.\nRemove it from configuration and try again.";
+							warnings.loc = node.loc;
+							module.addWarning(warnings);
+							return;
+						}
+
+						modulesMap.set(module, sourceMappingURL);
+					});
+				};
+
+				compilation.hooks.finishModules.tapAsync(
+					"ExtractSourceMapPlugin",
+					() => {
+						if (modulesMap.size === 0) return;
+						asyncLib.each(
+							modulesMap,
+							([module, sourceMappingURL], callback) => {
+								fetchFromURL(
+									compiler.inputFileSystem,
+									module.context || "",
+									sourceMappingURL,
+									"",
+									(err, result) => {
+										if (err) {
+											module.addWarning(err);
+											return callback();
+										}
+										const { sourceURL, sourceContent } = result;
+
+										if (module.buildInfo && module.buildInfo.fileDependencies) {
+											module.buildInfo.fileDependencies.add(sourceURL);
+										}
+
+										let map;
+
+										try {
+											map = JSON.parse(sourceContent.replace(/^\)\]\}'/, ""));
+										} catch (parseError) {
+											module.addWarning(
+												new WebpackError(
+													`Failed to parse source map from '${sourceMappingURL}': ${parseError}`
+												)
+											);
+											return callback();
+										}
+
+										// if (map.sections) {
+										// map = await flattenSourceMap(map);
+										// }
+										// add source content?
+										module.setSourceMap(map, compiler.context);
+										return callback();
+									}
+								);
+							}
+						);
+					}
+				);
+
+				normalModuleFactory.hooks.parser
+					.for("javascript/auto")
+					.tap("ExtractSourceMapPlugin", handler);
+				normalModuleFactory.hooks.parser
+					.for("javascript/esm")
+					.tap("ExtractSourceMapPlugin", handler);
+				normalModuleFactory.hooks.parser
+					.for("javascript/dynamic")
+					.tap("ExtractSourceMapPlugin", handler);
+			}
```

Comment:
There is a problem, `source-map-loader` supports any type of files (css/js/ts/etc), I think it's worth thinking about it. 

Also I think it should be a module option, i.e.

```js
module.exports = {
  //...
  module: {
    rules: [
      //...
      {
        test: /\.json$/,
        type: 'javascript/auto',
        // Maybe better name
        loadSourceMap: true,
        loader: 'custom-json-loader',
      },
    ],
  },
};
```

So you can manage where to consume and where not to, it's more flexible and I think it is right direction

Reply:
we can add this only to webpack supported module types ( in source map case it is `javascript/*` or `css/*`)

Reply:
also does `source-map-loader` works with loaders chain like => `source-map-loader` -> `babel-loader`. My question is should we "detect" source map after loaders run?

Reply:
I think we can add support to any type, we just need to load source map when `loadSourceMap` is true and pass them to loader, currently solution lose source map, loader can generate additional code and change source map

Reply:
>I think we can add support to any type, we just need to load source map when loadSourceMap is true and pass loaded source map to loader(s)

lets add this to parser option.

> loader can generate additional code and change source map

need @sokra opinion on that.. In this case I think we should keep loader solution. Core should support only source map extraction after loaders run. Since otherwise we still lose source map, if e.g. `babel-loader` will generate another source map.

Reply:
I also think that we should extract the source map before loaders run.

We can assume (or have to) that loaders handle source map themselves. (They get passed input source map and can return generated source map)

Do we can handle a source map when reading the resource. In NormalModule, where the resource is read.

This might require some changes in loader-runner, since they need to allow returning multiple arguments from `readResource`, or something like that.

We have to think about binary vs string representation of resources. We don't want to do Source Map extraction of binary resources, either do we want to unnecessary convert binary to string. So I guess we want to have a readResource callback where we know if the next loader expects binary or string.

---

(A loader generating a `sourceMappingURL` comment is an incorrect loader, since that's not the way loaders should return a source map. But in that case users can still "fix" the loader by adding the `source-map-loader`, but that's still an exception)

Reply:
I have refactored design:
- extract source map before loaders run ( while processing resource in `loader-runner` )
- try to extract source map **only** when option `parserOptions.extractSourceMap===true`
- when extracting explicitly convert resource content to string